browser/ui/webui/settings/brave_tor_handler.cc

```diff
@@ -0,0 +1,281 @@
+    DCHECK_EQ(State::kWaitForBridges, state_);
+    render_frame_host_->ExecuteJavaScriptInIsolatedWorld(
+        kParseBridgesScript,
+        base::BindOnce(&BridgeRequest::OnBridgesParsed,
+                       weak_factory_.GetWeakPtr()),
+        kIsolatedWorldId);
+  }
+
+  void OnBridgesParsed(base::Value value) {
+    std::move(result_callback_).Run(std::move(value));
+  }
+
+  raw_ptr<content::RenderFrameHost> render_frame_host_ = nullptr;
+  CaptchaCallback captcha_callback_;
+  BridgesCallback result_callback_;
+  State state_ = State::kLoadCaptcha;
+  std::unique_ptr<content::WebContents> web_contents_;
+  std::unique_ptr<image_fetcher::ImageDecoder> image_decoder_;
+
+  base::WeakPtrFactory<BridgeRequest> weak_factory_{this};
+};
+
+BraveTorHandler::BraveTorHandler() = default;
+BraveTorHandler::~BraveTorHandler() = default;
+
+void BraveTorHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "brave_tor.getBridgesConfig",
+      base::BindRepeating(&BraveTorHandler::GetBridgesConfig,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "brave_tor.setBridgesConfig",
+      base::BindRepeating(&BraveTorHandler::SetBridgesConfig,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "brave_tor.requestBridgesCaptcha",
+      base::BindRepeating(&BraveTorHandler::RequestBridgesCaptcha,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "brave_tor.resolveBridgesCaptcha",
+      base::BindRepeating(&BraveTorHandler::ResolveBridgesCaptcha,
+                          base::Unretained(this)));
+}
+
+void BraveTorHandler::GetBridgesConfig(const base::Value::List& args) {
+  AllowJavascript();
+
+  CHECK_EQ(1u, args.size());
+  const auto bridges_config = TorProfileServiceFactory::GetTorBridgesConfig();
+
+  ResolveJavascriptCallback(args[0], bridges_config.ToValue());
+}
+
+void BraveTorHandler::SetBridgesConfig(const base::Value::List& args) {
+  CHECK_EQ(1u, args.size());
+  CHECK(args[0].is_dict());
+
+  const auto bridges_config = tor::BridgesConfig::FromValue(&args[0]);
+  CHECK(bridges_config);
+  TorProfileServiceFactory::SetTorBridgesConfig(*bridges_config);
+}
+
+void BraveTorHandler::RequestBridgesCaptcha(const base::Value::List& args) {
+  auto captcha_callback =
+      base::BindOnce(&BraveTorHandler::SendResultToJavascript,
+                     base::Unretained(this), args[0].Clone());
+  request_ = std::make_unique<BridgeRequest>(
+      web_ui()->GetWebContents()->GetBrowserContext(),
+      std::move(captcha_callback));
+}
+
+void BraveTorHandler::ResolveBridgesCaptcha(const base::Value::List& args) {
+  AllowJavascript();
+
+  auto bridges_callback =
+      base::BindOnce(&BraveTorHandler::SendResultToJavascript,
+                     base::Unretained(this), args[0].Clone());
+  request_->ProvideCaptcha(args[1].GetString(), std::move(bridges_callback));
```

Comment:
We need to check `request_` is not nullptr before resolving captcha. Also we should invalidate `request_` afterward since the state machine in `BridgeRequest` can only allow one attempt to try, if that captcha resolving failed, we need to fetch new image hence new request. We can't let API caller to call two `ResolveBridgesCaptcha` consecutively

Reply:
Done.